README.md
# dialog4search

A package in which user can search for an item in a given list.

## Usage

To use this plugin, add dialog4search as a [dependency in your pubspec.yaml file.](https://flutter.dev/docs/development/packages-and-plugins/using-packages)

## Dialog with single selection

![Select Single](screenshots/singleSelect.gif)

## Dialog with multiple selection

![Select Multiple](screenshots/multipleSelect.gif)

### Example


```dart


import 'package:dialog4search/dialog4search.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Dialog4search',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Dialog4search Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;


  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Person> _selectedPersons=[];
  InputDecoration _searchBoxDecoration=InputDecoration(
      hintText: 'Search by father or sons name',
      contentPadding: EdgeInsets.symmetric(horizontal: 10,vertical: 0),
      border: OutlineInputBorder(borderSide: BorderSide(color: Colors.black)),
      filled: true,
      hintStyle: TextStyle(fontSize: 16, color: Colors.grey));

  bool _onSearch(person, value) {
    return person.fathersName
        .toUpperCase()
        .contains(value.toUpperCase()) ||
        person.name.toUpperCase().contains(value.toUpperCase());
  }

  Widget _itemBuilder(context, person) {
    return Container(
      color: Colors.white,
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text("Person: ${person.name}"),
          Text("Father: ${person.fathersName}"),
        ],
      ),
    );
  }

  Widget _builder(context, person) {
    return Card(
      color: Colors.blue,
      elevation: 5,
      child: Container(
        width: MediaQuery.of(context).size.width,
        height: 50,
        alignment: Alignment.centerLeft,
        padding: EdgeInsets.only(left: 16),
        child: Text("${person.name}",style: TextStyle(color: Colors.white,fontSize: 20),),
      ),
    );
  }

  void _onChange(persons){
    setState(() {
      _selectedPersons=persons;
    });
  }

  _displaySelectedItems(){
    return Expanded(child: ListView.separated(itemBuilder: (context,index){
      var person=_selectedPersons[index];
      return Container(
        color: Colors.white,
        padding: EdgeInsets.all(20),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text("Person: ${person.name}"),
            Text("Father: ${person.fathersName}"),
          ],
        ),
      );
    }, separatorBuilder: (context,index){
      return Divider();
    }, itemCount: _selectedPersons.length));
  }

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            SizedBox(height: 50,),
            Dialog4Search<Person>(
              //uncomment the line below to change the background color of the dialog box
              multipleSelect: true,
              searchFunction: _onSearch,
              itemBuilder: _itemBuilder,
              list: testData,
              builder: _builder,
              searchBoxDecoration: _searchBoxDecoration,
              onChanged: _onChange,
              onInitialization: (person){
                print("${person.name}");
              },
              initialValue: testData[2],

            //uncomment the line below to change the background color of the dialog box
//              dialogBoxBackgroundColor: Colors.red,

            // You can set your icon to show the election of your item.
//            selectionIcon: Icon(Icons.check),

            //Using dialog shape property you can change the shape of the dialog
//            dialogShape: ,
            //if you pass no match builder then it will be shown when item is not found in the list
//            noMatchBuilder: ,
            ),
            _displaySelectedItems()
          ],
        ),
      ),
    );
  }
}

class Person {
  final String name;
  final String fathersName;
  Person({this.name, this.fathersName});
}

var testData = [
  Person(name: 'Anuj Patel', fathersName: 'Ram Lal Patel'),
  Person(name: 'Anuj Verma', fathersName: 'Ram Lal Verma'),
  Person(name: 'Dinesh Mishra', fathersName: 'Mintu Mishra'),
  Person(name: 'Kapil Gupta', fathersName: 'Jhanni Lal Gupta'),
  Person(name: 'Rajpal verma', fathersName: 'Baburam verma'),
  Person(name: 'Satendra Rathod', fathersName: 'Kamlesh Rathod'),
  Person(name: 'Shuresh Kumar', fathersName: 'Tikaram verma'),
  Person(name: 'Omkar kashyap', fathersName: 'Khanna kashyap'),
  Person(name: 'Sachin kumar', fathersName: 'Omkar kashyap'),
  Person(name: 'Sudheer Pandey', fathersName: 'Anchal Pandey'),
  Person(name: 'Priya prakash', fathersName: 'Ratnesh Kumar'),
  Person(name: 'Shipra', fathersName: 'Jayram Rajan'),
  Person(name: 'Druv Tiwari', fathersName: 'Govind Tiwari'),
  Person(name: 'Karnesh Baghel', fathersName: 'Karanjeet Baghel'),
  Person(name: 'Rupesh Chauhan', fathersName: 'Rakesh Chauhan'),
  Person(name: 'Somesh Yadav', fathersName: 'Ganesh Yadav'),
  Person(name: 'Dashmesh Singh', fathersName: 'Tara Singh'),
  Person(name: 'Tapshi juneja', fathersName: 'Amar juneja'),
  Person(name: 'Nihal Jaiswal', fathersName: 'Akash'),
  Person(name: 'Fatima ', fathersName: 'Akbar Ali'),
];



```

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [0.0.1] - initial open source release.



test/dialog4search_test.dart
//import 'package:flutter_test/flutter_test.dart';
//
//import 'package:dialog4search/dialog4search.dart';
//
//void main() {
//  test('adds one to input values', () {
//    final calculator = Calculator();
//    expect(calculator.addOne(2), 3);
//    expect(calculator.addOne(-7), -6);
//    expect(calculator.addOne(0), 1);
//    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//  });
//}


lib/dialog4search.dart
library dialog4search;

import 'package:flutter/material.dart';

class Dialog4Search<T> extends StatefulWidget {
  //Using this function user will define search conditions.
  final bool Function(T, String) searchFunction;
  //List in which we are going to search items;
  final List<T> list;
  //tapping on this widget will open the dialog.
  final Widget Function(BuildContext, T) builder;
  //This builder is used to build the list items in the dialog.
  final Widget Function(BuildContext, T) itemBuilder;
  //sets the initial value.
  final T initialValue;
  final Function(T) onInitialization;
  final Function(List<T>) onChanged;
  final InputDecoration searchBoxDecoration;
  final BoxConstraints dialogBoxConstraint;
  final WidgetBuilder noMatchBuilder;
  final Color dialogBoxBackgroundColor;
  final ShapeBorder dialogShape;
  final bool multipleSelect;
  final Widget selectionIcon;
  Dialog4Search(
      {@required this.searchFunction,
      @required this.itemBuilder,
      @required this.list,
      @required this.builder,
      this.initialValue,
      this.onChanged,
      this.onInitialization,
      this.searchBoxDecoration,
      this.dialogShape,
      this.dialogBoxBackgroundColor,
      this.dialogBoxConstraint,
      this.noMatchBuilder,
      this.multipleSelect = false,
      this.selectionIcon})
      : assert(searchFunction != null),
        assert(itemBuilder != null),
        assert(builder != null),
        assert(list != null && list.length > 0);
  @override
  State<StatefulWidget> createState() {
    return _Dialog4SearchState<T>();
  }
}

class ListItem<T> {
  bool selected;
  T item;
  final int id;
  ListItem({this.item, this.selected, @required this.id});
}

class SelectionDialog<T> extends StatefulWidget {
  final bool Function(T, String) searchFunction;
  final List<T> list;
  final Widget Function(BuildContext context, T) itemBuilder;
  final InputDecoration searchBoxDecoration;
  final BoxConstraints dialogBoxConstraint;
  final WidgetBuilder noMatchBuilder;
  final Color dialogBoxBackgroundColor;
  final ShapeBorder dialogShape;
  final bool multipleSelect;
  final Function(List<T>) onChanged;
  final Widget selectionIcon;
  SelectionDialog(
      {@required this.list,
      @required this.searchFunction,
      @required this.itemBuilder,
      this.searchBoxDecoration,
      this.dialogBoxConstraint,
      this.noMatchBuilder,
      this.dialogShape,
      this.dialogBoxBackgroundColor,
      this.multipleSelect,
      this.onChanged,
      this.selectionIcon});
  @override
  State<StatefulWidget> createState() {
    return _SelectionDialog<T>();
  }
}

class _Dialog4SearchState<T> extends State<Dialog4Search<T>> {
  T _selectedItem;
  @override
  Widget build(BuildContext context) {
    return Material(
      child: InkWell(
        child: widget.builder(context, _selectedItem),
        onTap: () {
          showDialog<List<T>>(
              context: context,
              builder: (context) {
                return SelectionDialog<T>(
                  itemBuilder: widget.itemBuilder,
                  list: widget.list,
                  searchFunction: widget.searchFunction,
                  searchBoxDecoration: widget.searchBoxDecoration,
                  dialogBoxBackgroundColor: widget.dialogBoxBackgroundColor,
                  dialogBoxConstraint: widget.dialogBoxConstraint,
                  dialogShape: widget.dialogShape,
                  multipleSelect: widget.multipleSelect,
                  onChanged: widget.onChanged,
                  selectionIcon: widget.selectionIcon,
                );
              }).then((value) {
            if (value != null) {
              setState(() {
                _selectedItem = value[0];
              });
              if (widget.onChanged != null) widget.onChanged(value);
            }
          });
        },
      ),
    );
  }

  @override
  void initState() {
    _selectedItem = widget.initialValue ?? widget.list[0];
    if (widget.onInitialization != null) widget.onInitialization(_selectedItem);
    super.initState();
  }
}

class _SelectionDialog<T> extends State<SelectionDialog<T>> {
  List<ListItem<T>> _fliteredItems = [];
  List<ListItem<T>> _listItems;
  ValueNotifier<String> _searchPhrase = ValueNotifier("");
  @override
  Widget build(BuildContext context) {
    return SimpleDialog(
      backgroundColor: widget.dialogBoxBackgroundColor,
      shape: widget.dialogShape,
      titlePadding: EdgeInsets.zero,
      title: Column(
        crossAxisAlignment: CrossAxisAlignment.end,
        mainAxisSize: MainAxisSize.min,
        children: [
          IconButton(
              icon:
                  widget.multipleSelect ? Icon(Icons.check) : Icon(Icons.close),
              onPressed: () {
                if (widget.multipleSelect) {
                  var items = _listItems
                      .where((element) => element.selected)
                      .toList()
                      .map<T>((e) => e.item)
                      .toList();
                  Navigator.of(context).pop(items.length > 0 ? items : null);
                } else
                  Navigator.of(context).pop(null);
              }),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: TextField(
              decoration: widget.searchBoxDecoration ??
                  InputDecoration(hintText: 'Search...'),
              onChanged: (value) {
                _filter(value);
              },
            ),
          )
        ],
      ),
      children: [
        ValueListenableBuilder(
            valueListenable: _searchPhrase,
            builder: (context, value, _) {
              return ConstrainedBox(
                constraints: widget.dialogBoxConstraint ??
                    BoxConstraints(
                      maxHeight: MediaQuery.of(context).size.height * 0.6,
                    ),
                child: Container(
                  child: _fliteredItems.length > 0
                      ? ListView.builder(
                          itemCount: _fliteredItems.length,
                          itemBuilder: (context, index) {
                            var item = _fliteredItems[index];
                            if (widget.multipleSelect)
                              return Material(
                                color: widget.dialogBoxBackgroundColor ??
                                    Colors.white,
                                child: InkWell(
                                  child: Padding(
                                    padding: const EdgeInsets.symmetric(
                                        horizontal: 8, vertical: 5),
                                    child: Row(
                                      crossAxisAlignment:
                                          CrossAxisAlignment.center,
                                      children: [
                                        Expanded(
                                            child: widget.itemBuilder(
                                                context, item.item)),
                                        SizedBox(
                                          width: 15,
                                        ),
                                        Opacity(
                                          child: widget.selectionIcon ??
                                              SizedBox(
                                                height: 20,
                                                width: 20,
                                                child: Container(
                                                  color: Theme.of(context)
                                                      .primaryColor,
                                                  alignment: Alignment.center,
                                                  child: Icon(
                                                    Icons.check,
                                                    color: Theme.of(context)
                                                        .primaryIconTheme
                                                        .color,
                                                    size: 15,
                                                  ),
                                                ),
                                              ),
                                          opacity:
                                              _fliteredItems[index].selected
                                                  ? 1
                                                  : 0.1,
                                        ),
                                        SizedBox(
                                          height: 20,
                                        )
                                      ],
                                    ),
                                  ),
                                  onTap: () {
                                    var listItemIndex = _listItems.indexWhere(
                                            (element) =>
                                        element.id ==
                                            _fliteredItems[index].id);
                                    setState(() {
                                      _listItems[listItemIndex].selected =
                                      !_listItems[listItemIndex].selected;
                                    });
                                  },
                                ),
                              );
                            return SimpleDialogOption(
                              child: widget.itemBuilder(context, item.item),
                              onPressed: () {
                                _selectItem(item.item);
                              },
                            );
                          },
                        )
                      : Column(
                          crossAxisAlignment: CrossAxisAlignment.center,
                          mainAxisAlignment: MainAxisAlignment.center,
                          children: [
                            widget.noMatchBuilder == null
                                ? Container(
                                    height: 100,
                                    child: Text('No match!!'),
                                  )
                                : widget.noMatchBuilder,
                          ],
                        ),
                ),
              );
            })
      ],
    );
  }

  @override
  void initState() {
    int id = 0;
    _listItems = widget.list
        .map<ListItem<T>>(
            (e) => ListItem(item: e, selected: false, id: id += 1))
        .toList();
    _fliteredItems = _listItems;
    super.initState();
  }

  _filter(String value) {
    _searchPhrase.value = value;
    if (value.isNotEmpty)
      _fliteredItems = _listItems.where((element) {
        return widget.searchFunction(element.item, value);
      }).toList();
    else
      _fliteredItems = _listItems;
  }

  _selectItem(e) {
    Navigator.pop(context, <T>[e]);
  }
}


example/lib/main.dart
import 'package:dialog4search/dialog4search.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Dialog4search',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Dialog4search Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;


  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Person> _selectedPersons=[];
  InputDecoration _searchBoxDecoration=InputDecoration(
      hintText: 'Search by father or sons name',
      contentPadding: EdgeInsets.symmetric(horizontal: 10,vertical: 0),
      border: OutlineInputBorder(borderSide: BorderSide(color: Colors.black)),
      filled: true,
      hintStyle: TextStyle(fontSize: 16, color: Colors.grey));

  bool _onSearch(person, value) {
    return person.fathersName
        .toUpperCase()
        .contains(value.toUpperCase()) ||
        person.name.toUpperCase().contains(value.toUpperCase());
  }

  Widget _itemBuilder(context, person) {
    return Container(
      color: Colors.white,
      child: Column(
        mainAxisSize: MainAxisSize.min,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text("Person: ${person.name}"),
          Text("Father: ${person.fathersName}"),
        ],
      ),
    );
  }

  Widget _builder(context, person) {
    return Card(
      color: Colors.blue,
      elevation: 5,
      child: Container(
        width: MediaQuery.of(context).size.width,
        height: 50,
        alignment: Alignment.centerLeft,
        padding: EdgeInsets.only(left: 16),
        child: Text("${person.name}",style: TextStyle(color: Colors.white,fontSize: 20),),
      ),
    );
  }

  void _onChange(persons){
    setState(() {
      _selectedPersons=persons;
    });
  }

  _displaySelectedItems(){
    return Expanded(child: ListView.separated(itemBuilder: (context,index){
      var person=_selectedPersons[index];
      return Container(
        color: Colors.white,
        padding: EdgeInsets.all(20),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text("Person: ${person.name}"),
            Text("Father: ${person.fathersName}"),
          ],
        ),
      );
    }, separatorBuilder: (context,index){
      return Divider();
    }, itemCount: _selectedPersons.length));
  }

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            SizedBox(height: 50,),
            Dialog4Search<Person>(
              //uncomment the line below to change the background color of the dialog box
              multipleSelect: true,
              searchFunction: _onSearch,
              itemBuilder: _itemBuilder,
              list: testData,
              builder: _builder,
              searchBoxDecoration: _searchBoxDecoration,
              onChanged: _onChange,
              onInitialization: (person){
                print("${person.name}");
              },
              initialValue: testData[2],

            //uncomment the line below to change the background color of the dialog box
//              dialogBoxBackgroundColor: Colors.red,

            // You can set your icon to show the election of your item.
//            selectionIcon: Icon(Icons.check),

            //Using dialog shape property you can change the shape of the dialog
//            dialogShape: ,
            //if you pass no match builder then it will be shown when item is not found in the list
//            noMatchBuilder: ,
            ),
            _displaySelectedItems()
          ],
        ),
      ),
    );
  }
}

class Person {
  final String name;
  final String fathersName;
  Person({this.name, this.fathersName});
}

var testData = [
  Person(name: 'Anuj Patel', fathersName: 'Ram Lal Patel'),
  Person(name: 'Anuj Verma', fathersName: 'Ram Lal Verma'),
  Person(name: 'Dinesh Mishra', fathersName: 'Mintu Mishra'),
  Person(name: 'Kapil Gupta', fathersName: 'Jhanni Lal Gupta'),
  Person(name: 'Rajpal verma', fathersName: 'Baburam verma'),
  Person(name: 'Satendra Rathod', fathersName: 'Kamlesh Rathod'),
  Person(name: 'Shuresh Kumar', fathersName: 'Tikaram verma'),
  Person(name: 'Omkar kashyap', fathersName: 'Khanna kashyap'),
  Person(name: 'Sachin kumar', fathersName: 'Omkar kashyap'),
  Person(name: 'Sudheer Pandey', fathersName: 'Anchal Pandey'),
  Person(name: 'Priya prakash', fathersName: 'Ratnesh Kumar'),
  Person(name: 'Shipra', fathersName: 'Jayram Rajan'),
  Person(name: 'Druv Tiwari', fathersName: 'Govind Tiwari'),
  Person(name: 'Karnesh Baghel', fathersName: 'Karanjeet Baghel'),
  Person(name: 'Rupesh Chauhan', fathersName: 'Rakesh Chauhan'),
  Person(name: 'Somesh Yadav', fathersName: 'Ganesh Yadav'),
  Person(name: 'Dashmesh Singh', fathersName: 'Tara Singh'),
  Person(name: 'Tapshi juneja', fathersName: 'Amar juneja'),
  Person(name: 'Nihal Jaiswal', fathersName: 'Akash'),
  Person(name: 'Fatima ', fathersName: 'Akbar Ali'),
];
